src/index.tsx

```diff
@@ -537,29 +552,35 @@ class App extends React.Component<{}, AppState> {
               let isDraggingElements = false;
               const cursorStyle = document.documentElement.style.cursor;
               if (this.state.elementType === "selection") {
-                const hitElement = elements.find(element => {
+                const hitElementIdx = this.state.elements.findIndex(element => {
                   return hitTest(element, x, y);
                 });
 
                 // If we click on something
-                if (hitElement) {
-                  if (hitElement.isSelected) {
+                if (hitElementIdx > -1) {
+                  const auxElement = { ...this.state.elements[hitElementIdx] };
+                  if (auxElement.isSelected) {
                     // If that element is not already selected, do nothing,
                     // we're likely going to drag it
                   } else {
                     // We unselect every other elements unless shift is pressed
                     if (!e.shiftKey) {
-                      clearSelection();
+                      this.clearSelection();
                     }
                     // No matter what, we select it
-                    hitElement.isSelected = true;
+                    auxElement.isSelected = true;
+                    // I think we need a better way to do this but...
+                    const newElements = [...this.state.elements];
+                    newElements.splice(hitElementIdx, 1, auxElement);
+
+                    this.setState({ elements: newElements });
                   }
                 } else {
                   // If we don't click on anything, let's remove all the selected elements
-                  clearSelection();
+                  this.clearSelection();
                 }
 
-                isDraggingElements = elements.some(
+                isDraggingElements = this.state.elements.some(
```

Comment:
@vjeux the problem i'm currently having is here.. So we clear the selection above, but since setState is async, when the code reaches here, there are still elements with `isSelected = true`.

This code is kinda confusing, because we're declaring onMouseMove and onMouseUp inside the onMouseDown event, and then adding it to the window.eventListeners.

Maybe someone with more experience than me here could help me understand this code?

Reply:
Normally, you're not supposed to be able to know what the mouse is doing when outside of the browser window for obvious reasons. Now, there's --one-- case where you can: when you pressed the mouse down and then leave the window before putting it up. This lets you drag things outside the window and back.

Not having this behavior is unexpected for the user. It's also error prone as if the user mouse up outside the window, we wouldn't get the event and things would be in a wacky when he mouses back in.

In order to implement this behavior, we have to register for the mousemove handler on the window object --within-- a mousedown handler. This is why we have this weird logic here.

https://stackoverflow.com/a/1745382/232122

Now, the annoying side effect of this in our case we need to be very careful what variables we capture. If we capture a variable when the mouse is down, it will never be updated since we're not going through the normal react lifecycle. So we need to always use this.state.something rather than `const something = this.state.something`.

I'm also not really sure what's happening to `this` to be honest, since we have the old one.

This is the reason why I migrated from hooks back to class component. I'm not exactly sure how to fix it with hooks.

I hope this gives you some more context and hopefully you can get it working. Otherwise, we can avoid blocking the rests of the pull requests and try it again later?

Reply:
Yeah let's merge the other MR's.. This one will take alot more time than i antecipated. Looks like we have lots of race conditions when moving everything inside the state. Might aswell merge the other features